### PR TITLE
Fixed #19263 -- Fixed crash when filtering using __in and an empty QuerySet.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/19263

### CODE REVIEW

**Comment by timgraham** [comment]
@jarshwah, I tend to think this probably isn't the ideal fix. Any advice?

File: django/db/models/lookups.py [hunk 1]
```diff
@@ -282,6 +282,9 @@ def process_rhs(self, compiler, connection):
             return super(In, self).process_rhs(compiler, connection)
 
     def get_rhs_op(self, connection, rhs):
+        if rhs == "()":
```
Comment (jarshwah) [inline django/db/models/lookups.py]: This part doesn't feel right to me. This means that we didn't successfully catch all code paths through the `process_rhs` method. "direct values" look like they're successfully caught, it's just when the rhs is a queryset (or multiple-value-producing-expression) that we need to rely on the get_rhs_op.

Unfortunately, I can't think of a better way of handling this at the moment. Maybe @akaariai has an idea?

If this part goes ahead though, we should clearly document why it's necessary so as not to confuse future us.

Comment (akaariai) [inline django/db/models/lookups.py]: I'm not sure, but maybe we could hack something in the query/queryset itself to not produce (), and instead raise EmptyResultSet directly. That is, somewhere Django is generating the string '()' for Food.objects.all()[0:0], maybe that code could raise EmptyResultSet instead?

Comment (timgraham) [inline django/db/models/lookups.py]: Tried moving the exception into the SQLCompiler -- still not too sure about it.

**Comment by timgraham** [comment]
Josh, Anssi, any intuitions if the solution in the second commit is better than the original proposal?

**Comment by jarshwah** [comment]
Yes I think I prefer this a little better. It seems like an implementation detail, so pushing that down into the compiler makes more sense to me. Your approach also seems to follow the same plan that all previous patches on the ticket did, so I'm happy enough with it at this point.

:shipit:

**Pull request merged**